mlrun/feature_store/api.py

```diff
@@ -451,38 +451,53 @@ def _ingest_with_spark(
-
-    df = source.to_spark_df(spark)
-    if featureset.spec.graph and featureset.spec.graph.states:
-        df = run_spark_graph(df, featureset, namespace, spark)
-    infer_from_static_df(df, featureset, options=infer_options)
-
-    key_column = featureset.spec.entities[0].name
-    timestamp_key = featureset.spec.timestamp_key
-    if not targets:
-        if not featureset.spec.targets:
-            featureset.set_targets()
-        targets = featureset.spec.targets
-        targets = [get_target_driver(target, featureset) for target in targets]
-
-    for target in targets or []:
-        spark_options = target.get_spark_options(key_column, timestamp_key)
-        logger.info(f"writing to target {target.name}, spark options {spark_options}")
-        df.write.mode("overwrite").save(**spark_options)
-        target.set_resource(featureset)
-        target.update_resource_status("ready")
-
-    _post_ingestion(mlrun_context, featureset, spark)
+    try:
+        if spark is None or spark is True:
+            # create spark context
+            from pyspark.sql import SparkSession
+
+            if mlrun_context:
+                session_name = f"{mlrun_context.name}-{mlrun_context.uid}"
+            else:
+                session_name = (
+                    f"{featureset.metadata.project}-{featureset.metadata.name}"
+                )
+
+            spark = SparkSession.builder.appName(session_name).getOrCreate()
+
+        df = source.to_spark_df(spark)
+        if featureset.spec.graph and featureset.spec.graph.states:
+            df = run_spark_graph(df, featureset, namespace, spark)
+        infer_from_static_df(df, featureset, options=infer_options)
+
+        key_column = featureset.spec.entities[0].name
+        timestamp_key = featureset.spec.timestamp_key
+        if not targets:
+            if not featureset.spec.targets:
+                featureset.set_targets()
+            targets = featureset.spec.targets
+            targets = [get_target_driver(target, featureset) for target in targets]
+
+        for target in targets or []:
+            if target.path and urlparse(target.path).scheme == "":
+                if mlrun_context:
+                    mlrun_context.logger.error(
+                        "Paths for spark ingest must contain schema, i.e v3io, s3, az"
+                    )
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Paths for spark ingest must contain schema, i.e v3io, s3, az"
+                )
+            spark_options = target.get_spark_options(key_column, timestamp_key)
+            logger.info(
+                f"writing to target {target.name}, spark options {spark_options}"
+            )
+            df.write.mode("overwrite").save(**spark_options)
+            target.set_resource(featureset)
+            target.update_resource_status("ready")
+        _post_ingestion(mlrun_context, featureset, spark)
+    finally:
+        if spark:
```

Comment:
_post_ingestion stops the spark session too, this means calling it twice in a successful flow.
I think that you should either call _post_ingestion with a success/error indication from finally or remove the spark.stop from _post_ingestion.

Reply:
i'll remove spark.close() from _post_ingestion

Reply:
sorry for nagging but it seems that _post_ingestion is called from ingest as well (there's a chance that there won't be a spark context there but we have to be sure)

Reply:
I see that we cannot get to this point with spark_context so please ignore